docs/src/proposals/multi-writers.md

```diff
@@ -0,0 +1,188 @@
+# Enable Multi-Writers in Shred Insertion
+
+## Problem
+The current shred insertion is single-threaded which prevents us from fully
+utilizing the write bandwidth of our RocksDB backed BlockStore.
+
+The entire shred insertion is protected by an `insert_shreds_lock`, which
+ensures the atomicity of the read-modify-write operations on meta-data
+column families.
+
+## Opportunities
+RocksDB has an API called merge operator.  In a high level, it offers
+lockless read-modify-write operation where we only write the delta
+as a merge operand to the DB, and such merge operand will be used to
+compute the final value by merging it with its older value.
+
+For instance, suppose we want to read a value and write value+1 back
+to RocksDB.  Instead of obtaining a write lock, reading the value,
+performing value + 1, writing value + 1 back to the DB then release
+the write lock, we simply write `+1` back to the DB and let such `+1`
+being processed only at the read time or compaction time.  This allows
+us to update an entry without the need of acquiring a write lock.
+
+## Existing Solution
+A write lock is used to perform read-modify-write during the shred insertion,
+which prevents us from having multiple writers.
+
+## Proposed Solution
+To enable running multiple writers in parallel, we need to ensure the
+atomicity of each shred insertion and avoid data races.  To achieve this,
+both the shared in-memory structures and BlockStore access have to be
+thread-safe.
+
+For the shared in-memory structures such as BlockstoreInsertionMetrics,
+we need to use a thread-safe version or only update them all at once
+at the end of the function call to minimize the lock duration.
+
+For the BlockStore, we will utilize RocksDB's merge operator to perform
+read-modify-write operation to avoid data race between different writers.
+
+Specifically, the `insert_shred()` function touches the following column
+families:
+
+### DeadSlots
+DeadSlots stores [(slot, flag)](https://github.com/solana-labs/solana/blob/master/ledger/src/blockstore.rs#L1212)
```

Comment:
for links to github, can you please use permalinks instead of links to current master code (which is constantly changing)? for example here the link would be:
https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore.rs#L1212

Reply:
Thanks for pointing this out.  Done!